lib/cli/watch-run.js

```diff
@@ -35,17 +62,30 @@ module.exports = (mocha, {ui}, fileCollectParams) => {
     // killed by SIGINT which has portable number 2.
     process.exit(128 + 2);
   });
+};
 
-  const watchFiles = utils.files(process.cwd(), fileCollectParams.extension);
-  let runAgain = false;
+/**
+ * Create an object that allows you to rerun tests on the mocha
+ * instance. `beforeRun` is called everytime before `mocha.run()` is
+ * called.
+ *
+ * @param {Mocha} mocha - Mocha instance
+ * @param {function} beforeRun - Called just before `mocha.run()`
+ */
+const createRerunner = (mocha, beforeRun) => {
+  // Set to a `Runner` when mocha is running. Set to `null` when mocha is not
+  // running.
+  let runner = null;
+
+  let rerunScheduled = false;
 
-  const loadAndRun = () => {
+  const run = () => {
     try {
-      mocha.files = files;
-      runAgain = false;
+      beforeRun();
+      resetMocha(mocha);
```

Comment:
Even for the first run - on the `ready` event - we reset everything, right?

Reply:
Yes. This is not ideal but it avoids special casing the initial run.